x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/ip/IpPrefix.java

```diff
@@ -0,0 +1,191 @@
+        description = "Truncates an IP to a given prefix length.",
+        examples = @Example(file = "ip", tag = "ipPrefix")
+    )
+    public IpPrefix(
+        Source source,
+        @Param(
+            name = "ip",
+            type = { "ip" },
+            description = "IP address of type `ip` (both IPv4 and IPv6 are supported)."
+        ) Expression ipField,
+        @Param(
+            name = "prefixLengthV4",
+            type = { "integer" },
+            description = "Prefix length for IPv4 addresses."
+        ) Expression prefixLengthV4Field,
+        @Param(
+            name = "prefixLengthV6",
+            type = { "integer" },
+            description = "Prefix length for IPv6 addresses."
+        ) Expression prefixLengthV6Field
+    ) {
+        super(source, Arrays.asList(ipField, prefixLengthV4Field, prefixLengthV6Field));
+        this.ipField = ipField;
+        this.prefixLengthV4Field = prefixLengthV4Field;
+        this.prefixLengthV6Field = prefixLengthV6Field;
+    }
+
+    public static IpPrefix readFrom(PlanStreamInput in) throws IOException {
+        return new IpPrefix(in.readSource(), in.readExpression(), in.readExpression(), in.readExpression());
+    }
+
+    public static void writeTo(PlanStreamOutput out, IpPrefix ipPrefix) throws IOException {
+        out.writeSource(ipPrefix.source());
+        List<Expression> fields = ipPrefix.children();
+        assert fields.size() == 3;
+        out.writeExpression(fields.get(0));
+        out.writeExpression(fields.get(1));
+        out.writeExpression(fields.get(2));
+    }
+
+    public Expression ipField() {
+        return ipField;
+    }
+
+    public Expression prefixLengthV4Field() {
+        return prefixLengthV4Field;
+    }
+
+    public Expression prefixLengthV6Field() {
+        return prefixLengthV6Field;
+    }
+
+    @Override
+    public boolean foldable() {
+        return Expressions.foldable(children());
+    }
+
+    @Override
+    public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        var ipEvaluatorSupplier = toEvaluator.apply(ipField);
+        var prefixLengthV4EvaluatorSupplier = toEvaluator.apply(prefixLengthV4Field);
+        var prefixLengthV6EvaluatorSupplier = toEvaluator.apply(prefixLengthV6Field);
+
+        return new IpPrefixEvaluator.Factory(
+            source(),
+            ipEvaluatorSupplier,
+            prefixLengthV4EvaluatorSupplier,
+            prefixLengthV6EvaluatorSupplier,
+            context -> new BytesRef(new byte[16])
+        );
+    }
+
+    @Evaluator
+    static BytesRef process(
+        BytesRef ip,
+        int prefixLengthV4,
+        int prefixLengthV6,
+        @Fixed(includeInToString = false, build = true) BytesRef scratch
+    ) {
+        if (prefixLengthV4 < 0 || prefixLengthV4 > 32) {
+            throw new IllegalArgumentException("Prefix length v4 must be in range [0, 32], found " + prefixLengthV4);
```

Comment:
If you do `@Evaluator(warnExceptions= IllegalArgumentException)` then these will turn into a warning and a `null` value which seems nice. Better than aborting the query with an exception.

Reply:
Oh ok, aboslutely